File: src/AST.cpp
#include "AST.h"
#include "Visitor.h"
#include "Memory.h"
#include "VisitorType.h"
#include <iostream>

using namespace std;


ASTNode::ASTNode(const Loc & loc)
:m_loc(loc)
{
}

Stmt::Stmt(const Loc & loc)
:ASTNode(loc)
{
}

StmtInNestedBlock::StmtInNestedBlock(const Loc & loc)
:Stmt(loc)
{ 
}

Expr::Expr(const Loc & loc)
:ASTNode(loc)
{
}

Decl::Decl(const string & name, const Loc & loc)
:ASTNode(loc) , m_name(name)
{ 
}


Identifier::Identifier(const string & name, const Loc & loc)
:Expr(loc) , m_name(name)
{
}

Identifier::~Identifier()
{
}

shared_ptr<Type> Identifier::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Identifier::Accept(Visitor & v)
{
    return v.visit(this);
}

Number::Number(int value, const Loc & loc)
:Expr(loc) , m_value(value)
{ 
}

Number::~Number()
{
}

shared_ptr<Type> Number::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Number::Accept(Visitor & v)
{
    return v.visit(this);
}

Null::Null(const Loc & loc)
:Expr(loc)
{
}

Null::~Null()
{
}

shared_ptr<Type> Null::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Null::Accept(Visitor & v)
{
    return v.visit(this);
}

BinaryOp::BinaryOp(BinaryOperator * op, Expr * left, Expr * right,const Loc & loc)
:Expr(loc) , m_operator(op) , m_left(left) , m_right(right)
{

}

BinaryOp::~BinaryOp()
{
    delete m_left;
    delete m_right;
}

shared_ptr<Type> BinaryOp::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * BinaryOp::Accept(Visitor & v)
{
    return v.visit(this);
}

int Plus::Apply(int left, int right)
{
    return left + right;
}

int Minus::Apply(int left, int right)
{
    return left - right;
}

int Times::Apply(int left, int right)
{
    return left * right;
}

int Divide::Apply(int left, int right)
{
    return left / right;
}

int Equal::Apply(int left, int right)
{
    return left == right;
}

int GreaterThan::Apply(int left, int right)
{
    return left > right;
}




CallFuncExpr::CallFuncExpr(Expr * targetFun, const vector<Expr*> arg, const Loc & loc)
:Expr(loc) , m_targetFun(targetFun) , m_arg(arg)
{
}

CallFuncExpr::~CallFuncExpr()
{
    for (uint16_t i = 0; i < m_arg.size(); i++)
        delete m_arg[i];
    delete m_targetFun;
}

shared_ptr<Type> CallFuncExpr::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * CallFuncExpr::Accept(Visitor & v)
{
    return v.visit(this);
}

Input::Input(const Loc & loc)
:Expr(loc)
{
}

Input::~Input()
{
}

shared_ptr<Type> Input::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Input::Accept(Visitor & v)
{
    return v.visit(this);
}

Alloc::Alloc(Expr * expr, const Loc & loc)
:Expr(loc) , m_expr(expr)
{
}

Alloc::~Alloc()
{
    delete m_expr;
}

shared_ptr<Type> Alloc::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Alloc::Accept(Visitor & v)
{
    return v.visit(this);
}

VarRef::VarRef(Identifier * id, const Loc & loc)
:Expr(loc) , m_id(id)
{
}

VarRef::~VarRef()
{
    delete m_id;
}

shared_ptr<Type> VarRef::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * VarRef::Accept(Visitor & v)
{
    return v.visit(this);
}

Deref::Deref(Expr * pointer, const Loc & loc)
:Expr(loc) , m_pointer(pointer)
{
}

Deref::~Deref()
{
    delete m_pointer;
}

shared_ptr<Type> Deref::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Deref::Accept(Visitor & v)
{
    return v.visit(this);
}

RecordField::RecordField(const string & name, Expr * expr, const Loc & loc)
:ASTNode(loc) , m_name(name) , m_expr(expr)
{
}

RecordField::~RecordField()
{
    delete m_expr;
}

shared_ptr<Type> RecordField::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * RecordField::Accept(Visitor & v)
{
    return v.visit(this);
}

Record::Record(const vector<RecordField*> & fields,const Loc & loc)
:Expr(loc) , m_fields(fields)
{ 
}

Record::~Record()
{
    for (uint16_t i = 0; i < m_fields.size(); i++)
        delete m_fields[i];
}

shared_ptr<Type> Record::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Record::Accept(Visitor & v)
{
    return v.visit(this);
}


FieldAccess::FieldAccess(Expr * record, const string & field, const Loc & loc)
:Expr(loc) , m_record(record) , m_field(field)
{
}

FieldAccess::~FieldAccess()
{
    delete m_record;
}

shared_ptr<Type> FieldAccess::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * FieldAccess::Accept(Visitor & v)
{
    return v.visit(this);
}

AssignStmt::AssignStmt(Expr * left, Expr * right, const Loc & loc)
:StmtInNestedBlock(loc) , m_left(left) , m_right(right)
{
}

AssignStmt::~AssignStmt()
{
    delete m_left;
    delete m_right;
}

shared_ptr<Type> AssignStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * AssignStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

Block::Block(const Loc & loc)
:Stmt(loc)
{
}

NestedBlockStmt::NestedBlockStmt(const vector<StmtInNestedBlock*> & body, const Loc & loc)
:Block(loc) , m_body(body)
{
}

NestedBlockStmt::~NestedBlockStmt()
{
    for (uint16_t i = 0; i < m_body.size(); i++)
        delete m_body[i];
}

shared_ptr<Type> NestedBlockStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * NestedBlockStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

IdentifierDecl::IdentifierDecl(const string & name, const Loc & loc)
:Decl(name,loc)
{ 
}

IdentifierDecl::~IdentifierDecl()
{
}

shared_ptr<Type> IdentifierDecl::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * IdentifierDecl::Accept(Visitor & v)
{
    return v.visit(this);
}

VarStmt::VarStmt(const vector<IdentifierDecl*> & decls, const Loc & loc)
:Stmt(loc) , m_decls(decls)
{
}

VarStmt::~VarStmt()
{
    for (uint16_t i = 0; i < m_decls.size(); i++)
        delete m_decls[i];
}

shared_ptr<Type> VarStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * VarStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

ReturnStmt::ReturnStmt(Expr * expr, const Loc & loc)
:Stmt(loc) , m_expr(expr)
{
}

ReturnStmt::~ReturnStmt()
{
    delete m_expr;
}

shared_ptr<Type> ReturnStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * ReturnStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

FunBlockStmt::FunBlockStmt(const vector<VarStmt*> & vars,
                 const vector<StmtInNestedBlock*> & stmts, 
                 ReturnStmt * ret,const Loc & loc)
:Block(loc) , m_vars(vars) , m_stmts(stmts) , m_ret(ret)
{
}

FunBlockStmt::~FunBlockStmt()
{
    for (uint16_t i = 0; i < m_vars.size(); i++)
        delete m_vars[i];
    for (uint16_t i = 0; i < m_stmts.size(); i++)
        delete m_stmts[i];
    delete m_ret;
}

shared_ptr<Type> FunBlockStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * FunBlockStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

IfStmt::IfStmt(Expr * guard, NestedBlockStmt * thenBranch, 
           NestedBlockStmt * elseBranch,const Loc & loc)
:StmtInNestedBlock(loc) , m_guard(guard) , m_thenBranch(thenBranch) , m_elseBranch(elseBranch)
{ 
}

IfStmt::~IfStmt()
{
    delete m_guard;
    delete m_thenBranch;
    delete m_elseBranch;
}

shared_ptr<Type> IfStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * IfStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

WhileStmt::WhileStmt(Expr * guard, NestedBlockStmt * block, const Loc & loc)
:StmtInNestedBlock(loc) , m_guard(guard) , m_block(block)
{
}

WhileStmt::~WhileStmt()
{
    delete m_guard;
    delete m_block;
}

shared_ptr<Type> WhileStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * WhileStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

OutputStmt::OutputStmt(Expr * expr, const Loc & loc)
:StmtInNestedBlock(loc) , m_expr(expr)
{ 
}

OutputStmt::~OutputStmt()
{
    delete m_expr;
}

shared_ptr<Type> OutputStmt::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * OutputStmt::Accept(Visitor & v)
{
    return v.visit(this);
}

FunDecl::FunDecl(const vector<IdentifierDecl*> params, FunBlockStmt * block,
            const string & name, const Loc & loc)
:Decl(name,loc) , m_params(params) , m_block(block)
{
}

FunDecl::~FunDecl()
{
    for (uint16_t i = 0; i < m_params.size(); i++)
        delete m_params[i];
    delete m_block;
}

shared_ptr<Type> FunDecl::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * FunDecl::Accept(Visitor & v)
{
    return v.visit(this);
}

Program::Program(vector<FunDecl*> & funs,const Loc & loc)
:ASTNode(loc) , m_funs(funs)
{
}

Program::~Program()
{
    for (uint16_t i = 0; i < m_funs.size(); i++)
        delete m_funs[i];
}

shared_ptr<Type> Program::Accept(VisitorType & v)
{
    return v.visit(this);
}

Object * Program::Accept(Visitor & v)
{
    return v.visit(this);
}

File: microc/src/main/scala/microc/ast/Ast.scala
package microc.ast

/**
  * A source code location of an AST node.
  *
  * @param line The line number.
  * @param col  The column number.
  */
case class Loc(line: Int, col: Int) extends Ordered[Loc] {
  override def toString: String = s"$line:$col"

  override def compare(that: Loc): Int = {
    val d = this.line - that.line
    if (d == 0) this.col - that.col else d
  }
}

/** A binary operator */
sealed trait BinaryOperator

object BinaryOperator {
  def apply(s: String): BinaryOperator = s match {
    case "+"  => Plus
    case "-"  => Minus
    case "*"  => Times
    case "/"  => Divide
    case "==" => Equal
    case ">"  => GreatThan
  }
}

case object Plus extends BinaryOperator {
  override def toString: String = "+"
}

case object Minus extends BinaryOperator {
  override def toString: String = "-"
}

case object Times extends BinaryOperator {
  override def toString: String = "*"
}

case object Divide extends BinaryOperator {
  override def toString: String = "/"
}

case object Equal extends BinaryOperator {
  override def toString: String = "=="
}

case object GreatThan extends BinaryOperator {
  override def toString: String = ">"
}

// ----------------------------------------------------------------------------
// BASE NODES
// ----------------------------------------------------------------------------

object AstNode {
  private val Printer = new AstPrinter()
}

/**
  * An AST node
  */
sealed abstract class AstNode {

  /**
    * The source code location of the AST node.
    *
    * @return a source code location.
    */
  def loc: Loc

  def children: Iterable[AstNode] = Iterable.empty

  def tree: Iterable[AstNode] = Iterable(this) ++ children.flatMap(_.tree)

  override def toString: String = {
    AstNode.Printer.print(this) + s"[$loc]"
  }
}

sealed trait Expr extends AstNode

sealed trait Stmt extends AstNode

sealed trait Decl extends AstNode {
  def name: String
}

/**
  * A statement in a nested block.
  *
  * It cannot be a declaration or a return.
  */
sealed trait StmtInNestedBlock extends Stmt

// ----------------------------------------------------------------------------
// EXPRESSIONS
// ----------------------------------------------------------------------------

case class Null(loc: Loc) extends Expr

case class Number(value: Int, loc: Loc) extends Expr

case class Identifier(name: String, loc: Loc) extends Expr

case class BinaryOp(operator: BinaryOperator, left: Expr, right: Expr, loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = List(left, right)
}

case class CallFuncExpr(targetFun: Expr, args: List[Expr], loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = targetFun :: args
}

/**
  * Read of one integer from standard input.
  *
  * @param loc The source code location.
  */
case class Input(loc: Loc) extends Expr

case class Alloc(expr: Expr, loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = List(expr)
}

/**
  * Variable reference.
  *
  * It is used to create a pointer to a variable `id` (&id).
  *
  * @param id  The target variable.
  * @param loc The source code location.
  */
case class VarRef(id: Identifier, loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = List(id)
}

/**
  * Pointer dereference
  *
  * It is used to dereference a pointer `p` (`*p`).
  *
  * @param pointer The pointer to dereference
  * @param loc     The source code location
  */
case class Deref(pointer: Expr, loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = List(pointer)
}

case class Record(fields: List[RecordField], loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = fields
}

case class RecordField(name: String, expr: Expr, loc: Loc) extends AstNode {
  override def children: Iterable[AstNode] = List(expr)
}

/**
  * Accessing a field of a record (x.y).
  *
  * @param record The expression that evaluates to a record value.
  * @param field  The name of the field of the record.
  * @param loc    The source code location.
  */
case class FieldAccess(record: Expr, field: String, loc: Loc) extends Expr {
  override def children: Iterable[AstNode] = List(record)
}

// ----------------------------------------------------------------------------
// STATEMENTS
// ----------------------------------------------------------------------------

/**
  * An Assignment.
  *
  * @param left  The LValue (where to assign)
  * @param right The RValue (what to assign)
  * @param loc   The source code location
  */
case class AssignStmt(left: Expr, right: Expr, loc: Loc) extends StmtInNestedBlock {
  override def children: Iterable[AstNode] = List(left, right)
}

case object DirectWrite {
  def unapply(expr: Expr): Option[Identifier] = expr match {
    case x: Identifier => Some(x)
    case _ => None
  }
}

case object IndirectWrite {
  def unapply(expr: Expr): Option[Deref] = expr match {
    case x: Deref => Some(x)
    case _ => None
  }
}

case object DirectFieldWrite {
  def unapply(expr: Expr): Option[(Identifier, String, Loc)] = expr match {
    case FieldAccess(record: Identifier, field, loc) => Some((record, field, loc))
    case _                                           => None
  }
}

case object IndirectFieldWrite {
  def unapply(expr: Expr): Option[(Deref, String, Loc)] = expr match {
    case FieldAccess(record: Deref, field, loc) => Some((record, field, loc))
    case _ => None
  }
}

sealed trait Block extends Stmt {
  def body: List[Stmt]

  override def children: Iterable[AstNode] = body
}

object Block {
  def unapply(that: Block): Option[List[Stmt]] = Some(that.body)
}

/**
  * Represents a block of statements surrounded by curly braces.
  *
  * {{{
  * {
  *   stmt_1
  *   stmt_2
  *   ...
  *   stmt_n
  * }
  * }}}
  *
  * @param body The list of statements in this block
  * @param loc The location
  */
case class NestedBlockStmt(body: List[StmtInNestedBlock], loc: Loc) extends Block with StmtInNestedBlock

case class FunBlockStmt(vars: List[VarStmt], stmts: List[StmtInNestedBlock], ret: ReturnStmt, loc: Loc) extends Block {
  val body: List[Stmt] = vars ++ (stmts :+ ret)
}

case class ReturnStmt(expr: Expr, loc: Loc) extends Stmt {
  override def children: Iterable[AstNode] = List(expr)
}

case class IfStmt(guard: Expr, thenBranch: StmtInNestedBlock, elseBranch: Option[StmtInNestedBlock], loc: Loc)
    extends StmtInNestedBlock {
  override def children: Iterable[AstNode] = guard :: thenBranch :: elseBranch.map(x => List(x)).getOrElse(Nil)
}

case class WhileStmt(guard: Expr, block: StmtInNestedBlock, loc: Loc) extends StmtInNestedBlock {
  override def children: Iterable[AstNode] = List(guard, block)
}

case class OutputStmt(expr: Expr, loc: Loc) extends StmtInNestedBlock {
  override def children: Iterable[AstNode] = List(expr)
}

/**
  * Function local variables declaration.
  *
  * @param decls The list of variable declarations.
  * @param loc   The source code location.
  */
case class VarStmt(decls: List[IdentifierDecl], loc: Loc) extends Stmt {
  override def children: Iterable[AstNode] = decls
}

case class IdentifierDecl(name: String, loc: Loc) extends Decl

case class FunDecl(name: String, params: List[IdentifierDecl], block: FunBlockStmt, loc: Loc) extends Decl {
  override def toString: String = s"$name(${params.mkString(",")}){...}:$loc"

  override def children: Iterable[AstNode] = params :+ block
}

/**
  * The complete program.
  *
  * A program is just a list of functions.
  * The `main` function is where the execution begins.
  *
  * @param funs The list of functions.
  * @param loc  The source code location.
  */
case class Program(funs: List[FunDecl], loc: Loc) extends AstNode {
  def mainFunOption: Option[FunDecl] =
    funs.find(_.name == "main")

  override def children: Iterable[AstNode] = funs

  def declarations: Iterable[Decl] = tree.collect { case x: Decl => x }
}
